7.0 sales analysis reporting in base currency

llacroix opened the pull request.

llacroix committed:
[#63600] Init module convert and save price in base currency

llacroix committed:
[#63600] Added report

llacroix committed:
[#63600] Split the module in 3 modules

sales_analysis_reporting_in_base_currency
- sales_analysis_reporting_in_base_currency_inherited
- sales_analysis_converting_sale_order_line_in_base_currency

llacroix committed:
[#63600] Fix names and errors for confirming lines

llacroix committed:
[#63600] Add predicate tests we can't always get the value

llacroix committed:
[#63600] Prevent some fields from being added as default if they aren't valid

llacroix committed:
[#68603] Added converted total to form view

llacroix committed:
[#68603] Added support to non editable order_lines.

The editor open a form view. We have to add an invisible field to
catch the on change events.

Also added a total converted to the form view.

llacroix committed:
Added converted subtotal one sale order lines and fixed somme issues

It might not be a good idea to store the fields yet. Stored fields are
computed only once and have to be recalculated.

llacroix committed:
Add context to methods

joaoalf commented:
@llacroix Can you please add some tests?

llacroix commented:
ok

llacroix committed:
Added tests that check if reports are being generated and have the field

llacroix committed:
Added more test to check conversion

llacroix committed:
Flake8 errors

joaoalf commented on sales_analysis_reporting_in_base_currency_inherited/report/sale_report.py:
```diff
@@ -0,0 +1,89 @@
+from openerp import tools
```
Copyright?

joaoalf commented on sales_analysis_reporting_in_base_currency_inherited/report/__init__.py:
```diff
@@ -0,0 +1 @@
+from . import sale_report
```
Copyright?

joaoalf commented on sales_analysis_reporting_in_base_currency_inherited/tests/test_report.py:
```diff
@@ -0,0 +1,50 @@
+from openerp.tests import common
```
Copyright?

joaoalf commented on sales_analysis_reporting_in_base_currency/__openerp__.py:
```diff
@@ -0,0 +1,72 @@
+
+This module has dependencies to :
+---------------------------------
+* account
```
@llacroix I guess it's better to document direct dependencies. If you put account as dependency in the doc but doesn't declare it as a dependency in the 'depends' section of the manifest many people will get confused.

joaoalf commented on sales_analysis_converting_sale_order_line_in_base_currency/tests/test_report.py:
```diff
@@ -0,0 +1,110 @@
+from openerp.tests import common
```
@llacroix Copyright?

joaoalf commented on sales_analysis_converting_sale_order_line_in_base_currency/tests/__init__.py:
```diff
@@ -0,0 +1,5 @@
+from . import test_report
```
Copyright?

joaoalf commented on sales_analysis_converting_sale_order_line_in_base_currency/__openerp__.py:
```diff
@@ -0,0 +1,73 @@
+This module has dependencies to :
+---------------------------------
+* account
+
```
@llacroix maybe you could add sale too.

llacroix committed:
Improved tests and review from PR

llacroix committed:
Added more copyright

llacroix committed:
Fix dependencies in openerp.py

llacroix committed:
Fix PEP8 errors

llacroix committed:
Added total with currencies to tree view

joaoalf commented on sales_analysis_converting_sale_order_line_in_base_currency/models/sale_order.py:
```diff
@@ -0,0 +1,281 @@
+    _inherit = 'sale.order.line'
+
+    _columns = {
+        "order_line_currency": fields.many2one(
```
@llacroix It's not mandatory but I would go with the name currency_id. We know that we are in the order line so the `order-line_;` prefix is not that relevant and  since it's many2one it's more clear to have the `_id` suffix.

llacroix replied:
Alright. I'm good with currency_id too

joaoalf commented on sales_analysis_converting_sale_order_line_in_base_currency/models/sale_order.py:
```diff
@@ -0,0 +1,281 @@
+        "order_line_currency": fields.many2one(
+            'res.currency', 'Currency'
+        ),
+        "amount_currency_calculated": fields.float(
```
@llacroix Can you add the precision?

llacroix replied:
ok

llacroix committed:
Splitted models/sale_order.py and added precision

llacroix committed:
Renamed order_line_currency to currency_id

llacroix committed:
Round values with precision in tests

llacroix committed:
In case lst_price is None use the line_price

llacroix commented:
:smile:

llacroix committed:
Added more tests

llacroix committed:
Added more tests

llacroix committed:
Test what happens when no currency is sent for lines and if price
matches the original price

llacroix committed:
Cast division to float for python2 integer division

llacroix commented:
Maxed the coverage.

llacroix committed:
Use the "Price Unit" as precision.

llacroix committed:
Changed the label of "Converted Amount" and removed Company currency form view

llacroix committed:
Fix roundings

llacroix committed:
Sum rounded subtotals instead of rounding only the final total

Even if the rounding for subtotal is set to two while price unit is 5,
OpenERP will keep 5 digits everywhere until it gets to the database or
the view. For this reason, we have to do roundings by ourselves in some
places.

In this case, OpenERP seems to be summing the rounded subtotals instead
of doing a sum on all lines and then rounding.

llacroix committed:
Fixed PEP8 error

eLBati commented:
Hi @llacroix 
as module description says

> This total is wrong
> if you have made sales in different currencies as
> Odoo ignores currencies in this instance

Shouldn't this be fixed in the core module?

gurneyalex commented on sales_analysis_converting_sale_order_line_in_base_currency/__openerp__.py:
```diff
@@ -0,0 +1,74 @@
+{
+    'name': 'sales Analysis converting sale order line in base currency',
+    'version': '0.1',
+    'author': 'Savoir-faire Linux',
```
add Odoo Community Association (OCA)

llacroix commented:
@eLBati It's been a long time... Obviously it should be fixed in the core modules but that would also mean that queries would be probably much more complicated than "fetch" and "show".

eLBati commented:
@llacroix I made something similar for invoice analysis: https://github.com/odoo/odoo/pull/7550
Would that approach be followable/correct?

pedrobaeza commented:
This is very old and without further activity, so I'm closing. If you are still interested in this, feel free to reopen it.
